src/Utilities/GenerateInstantiations.hpp

```diff
@@ -0,0 +1,165 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/preprocessor/list/for_each_product.hpp>
+#include <boost/preprocessor/list/size.hpp>
+#include <boost/preprocessor/list/to_tuple.hpp>
+#include <boost/preprocessor/list/transform.hpp>
+#include <boost/preprocessor/tuple/elem.hpp>
+#include <boost/preprocessor/tuple/size.hpp>
+#include <boost/preprocessor/tuple/to_list.hpp>
+#include <boost/preprocessor/variadic/to_list.hpp>
+
+/// \cond
+#define GENERATE_INSTANTIATIONS_DO_PRODUCT(INSTANTIATION_MACRO, LIST_OF_LISTS) \
+  BOOST_PP_LIST_FOR_EACH_PRODUCT(INSTANTIATION_MACRO,                          \
+                                 BOOST_PP_LIST_SIZE(LIST_OF_LISTS),            \
+                                 BOOST_PP_LIST_TO_TUPLE(LIST_OF_LISTS))
+
+#define GENERATE_INSTANTIATION_TUPLES_TO_LISTS(d, _, elem) \
+  BOOST_PP_TUPLE_TO_LIST(BOOST_PP_TUPLE_SIZE(elem), elem)
+/// \endcond
+
+/*!
+ * \ingroup Utilities
+ * \brief Macro useful for generating many explicit instantiations of function
+ * or class templates
+ *
+ * It is often necessary to generate explicit instantiations of function or
+ * class templates. Since the total number of explicit instantiation scales the
+ * product of the number of possible number of parameter values of each template
+ * parameter, this quickly becomes tedious. This macro allows you to easily
+ * generate hundreds of explicit instantiations.
+ *
+ * The first argument to the macro is a macro that takes two argument described
+ * below. The remaining arguments are macro-tuples, e.g. `(1, 2, 3)`. The
+ * Cartesian product of the macro-tuples is then computed and each term is
+ * passed as a tuple as the second argument to the `INSTANTIATION_MACRO`. The
+ * first argument to the `INSTANTIATION_MACRO` in a Boost.Preprocessor internal
```

Comment:
this sentence should be by the macro that uses the `_` as I find it to be confusing here